ulib/LowStar.Literal.fsti

```diff
@@ -0,0 +1,137 @@
+  Seq.seq_of_list cs
+
+/// Consequently, this function becomes in C a simple cast from ``const char *`` to
+/// ``char *``, since immutable buffers don't (yet) have the ``const`` attribute in
+/// KreMLin. (This is unsavory, and should be fixed later.) This way, a string
+/// literal can be seen as an immutable buffer and passed around as such.
+/// This function checks at extraction-time that its argument is a literal.
+val buffer_of_literal: (s: ascii_string) ->
+  ST.Stack (b: IB.ibuffer UInt8.t)
+    (requires (fun _ ->
+      // TODO: find a way to meta-program the "is literal" check
+      String.length s > 0 /\ // needed because of the precondition on igcmalloc_of_list
+      String.length s < pow2 32))
+    (ensures (fun h0 b h1 ->
+      MB.frameOf b == HS.root /\ // is this really useful?
+      MB.recallable b /\
+      MB.alloc_post_mem_common b h0 h1 (u8s_of_ascii_string s)))
+
+/// .. note::
+///
+///    This literal will be zero-terminated, but since we do not require that the
+///    string literal be zero-free, the trailing zero will be ignored and unused. This
+///    means that we won't be able to use the C standard library functions for string
+///    manipulation and will instead have to pass lengths at run-time.
+
+/// Rather than having to write ``assert_norm`` by hand, this convenient wrapper
+/// relies on the normalizer to discharge all the relevant proof obligations, and
+/// synthesizes the length of the resulting buffer. The pair has no cost: KreMLin
+/// guarantees that it will be eliminated.
+unfold
+let buf_len_of_literal (s: string):
+  ST.Stack (IB.ibuffer UInt8.t & UInt32.t)
+    (requires (fun _ ->
+      normalize (String.length s > 0) /\
+      normalize (is_ascii_string s) /\
+      normalize (List.Tot.length (String.list_of_string s) < pow2 32)))
+    (ensures (fun h0 r h1 ->
+      let b, l = r in
+      MB.frameOf b == HS.root /\
+      MB.recallable b /\
+      MB.alloc_post_mem_common b h0 h1 (u8s_of_ascii_string s) /\
+      UInt32.v l = IB.length b))
+=
+  [@inline_let]
+  let l = normalize_term (UInt32.uint_to_t (List.Tot.length (String.list_of_string s))) in
+  buffer_of_literal s, l
+
+/// String literals as values
+/// -------------------------
+///
+/// The other way to leverage string literals in Low* is to see them as values
+/// (they are always live, because they live in the data segment). This enables
+/// a variety of use-cases, mostly for printing functions. For these, we
+/// currently do not enable any sort of advanced reasoning, except that they must
+/// not contain zeroes (since printing routines rely on the trailing zero). Note
+/// that we don't reflect the fact that these are zero-terminated like ``C.String``
+/// attempted to do.
+
+let zero_free s =
+  List.Tot.for_all (fun c -> Char.int_of_char c <> 0) (String.list_of_string s)
+
+val const_string: Type0
+
+val const_string_of_literal: s:string { normalize (zero_free s) } -> const_string
```

Comment:
How will the clients prove the precondition? `String.list_of_string` is a `val` in `FStar.String.fsti`, and there don't seem to be any lemmas around it?